### DISCUSSION TITLE
How to override the default value errors on pydantic validation errors of the responses

### DISCUSSION THREAD

**Discussion by hhertogh** [fastapi/fastapi#6048] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class DiscountItem(BaseModel):
    name: Optional[str] = None
    total: Decimal

@router.get(
    "/sales",
    response_model=DiscountItem,
    summary="Get sales")
async def get_sales(request: Request):
    return {}
```

### Description

For the example above, you will get a 500 internal server error and on the server you will see ValidationError and value_error.missing

Is there a way to override those errors ? 
For example, catch the errors , do something and then raise the errors again or send a customized error back to the customers ?

I've seen this [example](https://stackoverflow.com/questions/70167626/how-to-prevent-pydantic-from-throwing-an-exception-on-validationerror) that provides a BaseModelNoException but it looks quite complex. Is there an easier way to achieve this ? I would like to catch the error and have access to the request object.

### Operating System

macOS

### FastAPI Version

0.86.0

### Python Version

3.10

**Comment by aybruhm** [comment]
The error messages you are getting comes from fastapi.exceptions, here's a [solution](https://stackoverflow.com/questions/60274141/is-it-possible-to-change-the-pydantic-error-messages-in-fastapi#60274832) I found on stackoverflow that solves the problem you are facing!

**Comment by jayashankarjayan** [comment, marked as answer]
@hhertogh , could implement a custom exception handler as shown in this section of the [documentation](https://fastapi.tiangolo.com/tutorial/handling-errors/?h=error).

I have taken your use case and extended it using a custom exception handler. The code for the same is as follows

```
from decimal import Decimal
from typing import Optional, List

from fastapi import (
    FastAPI, APIRouter, Request,
    status
    )
from fastapi.responses import JSONResponse
from pydantic import BaseModel
from pydantic.error_wrappers import ValidationError


class DiscountItem(BaseModel):
    name: Optional[str] = None
    total: Decimal

class BasicResponse(BaseModel):
    message: List[str]
    status: bool

app = FastAPI()

@app.exception_handler(ValidationError)
async def validation_exception_handler(request, exc):
    messages = []
    for error in exc.errors():
        message = f"{error['msg']}: {error['loc'][-1]}"
        messages.append(message)
    response = BasicResponse(message=messages, status=False).dict()
    
    return JSONResponse(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                         content=response)


router = APIRouter()

@router.get(
    "/sales",
    response_model=DiscountItem,
    summary="Get sales")
async def get_sales(request: Request):
    return {}

app.include_router(router)
```

**Comment by hhertogh** [comment]
Thanks @jayashankarjayan. This works.

I've done other tests on validation of inputs as well.
If you have for example a field missing in the request, it will raise a RequestValidationError (fastapi.exceptions).
If you add a validation on the input for example a min_length of 1 (you don't want your input to be empty), then it will raise a ValidationError, as follows:
 ```
[{'loc': ('response', 'name'), 'msg': 'ensure this value has at least 1 characters', 'type': 'value_error.any_str.min_length', 'ctx': {'limit_value': 1}}]
```
Why is not raising a requestvalidationerror ? Because it's a pydantic and not fastapi error ?

Thanks for the help!

**Comment by hhertogh** [comment]
Another question would be: is it possible to override the default timeout error ?
We have for example a timeout of 60 sec managed by gunicorn, but I would like to send a proper timeout response to my users and log the timeout in my db.
Thanks.
